packages/protocol/test/governance/validators.ts

```diff
@@ -2178,10 +2178,98 @@ contract('Validators', (accounts: string[]) => {
       })
     })
 
-    describe('when the sender is not an approved address', async () => {
+    describe('when the sender is not an approved address', () => {
       it('should revert', async () => {
         await assertRevert(validators.forceDeaffiliateIfValidator(validator))
       })
     })
   })
+
+  describe('#groupMembershipInEpoch', () => {
+    const validator = accounts[0]
+    const groups = accounts.slice(1, -1)
+
+    describe('when the validator is added to different groups with gaps in between epochs', () => {
+      let epochs = []
+
+      beforeEach(async () => {
+        // epochs stores [epochNumber, group] of the corresponding i + 1 indexed entry on chain
+        // i + 1 because registering groups adds a dummy null address as the first entry
+        epochs = []
+        await registerValidator(validator)
+        for (const group of groups) {
+          await registerValidatorGroup(group)
+        }
+        // Magic gapSize to allow for tests of off by one in both directions
+        const gapSize = 3
+        for (let i = 0; i < membershipHistoryLength.multipliedBy(gapSize).toNumber(); i++) {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          const blocksUntilNextEpoch = (epochNumber + 1) * EPOCH - blockNumber
+          await mineBlocks(blocksUntilNextEpoch, web3)
+
+          if (i % gapSize === 0) {
+            // Current epochNumber is 1 greater since we just called `mineBlocks`
+            // We set the 4th (3 index) entry to 0x00 to test possible initialization bugs
+            epochs.push([epochNumber + 1, i !== gapSize * 3 ? groups[i / gapSize] : NULL_ADDRESS])
+            // If not 4th entry (0 indexed)
+            if (i !== gapSize * 3) {
+              await validators.affiliate(groups[i / gapSize])
+              await validators.addFirstMember(validator, NULL_ADDRESS, NULL_ADDRESS, {
+                from: groups[i / gapSize],
+              })
+            } else {
+              await validators.deaffiliate()
+            }
+          }
+        }
+      })
+
+      it('should correctly get the group address for exact epoch numbers', async () => {
+        for (let i = 0; i < membershipHistoryLength.toNumber(); i++) {
+          assert.equal(
+            await validators.groupMembershipInEpoch(validator, epochs[i][0], 1 + i),
+            i !== 3 ? epochs[i][1] : NULL_ADDRESS
+          )
+        }
+      })
+
+      it('should correctly get the group address of epoch numbers between switches', async () => {
+        assert.equal(
```

Comment:
Why not test these conditions in a for loop through over all epochs? As is I find it a bit hard to read